File: Mqtt.razor.js
﻿class Mqtt
{
	/**
	 * 如果没有传递参数，则会使用匿名连接
	 * @param {string} username
	 * @param {string} password
	 */
	constructor(dotnetHelper, username, password, clientId)
	{
		let options = {
			// Clean session
			clean: true,
			connectTimeout: 3000,
			keepalive: 15,
			// Auth
			username,
			password,
			clientId,
		}
		this.dotnetHelper = dotnetHelper;
		this.client = mqtt.connect("ws://127.0.0.1:8083/mqtt", options);
		this.b_HadBeenConnected = false;//过去曾经连接成功过
		this.client.on("reconnect", () =>
		{
			/**
			 * 请求被emqx拒绝后并不会进入 error 事件，而是不断进入
			 * reconnect 事件
			 * 没办法，只能通过这个事件来通知用户连接失败。为了区分到底是连接失
			 * 败还是短线重连，需要设置一个标志位 b_HadBeenConnected
			 */
			if (!this.b_HadBeenConnected)
			{
				//触发 .NET 的连接失败事件
				this.dotnetHelper.invokeMethodAsync("OnConnectFail");
				this.client.end(true);//关闭客户端，防止一直重连
			}
			else
			{
				console.log("重连");
			}
		});
		this.client.on("connect", () =>
		{
			this.b_HadBeenConnected = true;
			this.dotnetHelper.invokeMethodAsync("OnConnected");
		});
		//在收到数据后将数据传给.NET方法
		this.client.on("message", (topic, payload) =>
		{
			payload = Uint8Array.from(payload.toJSON().data);
			this.dotnetHelper.invokeMethodAsync("OnReceived", topic, payload);
		});
	}
	/**
	 * 发布主题
	 * @param {string} topic
	 * @param {Uint8Array} payload
	 */
	publish(topic, payload)
	{
		this.client.publish(topic, payload, { qos: 1 });
	};
	/**
	 * 订阅主题
	 * @param {any} topic
	 */
	subscribe(topic)
	{
		this.client.subscribe(topic, { qos: 1 }, (error, granted) =>
		{
			if (error)
			{
				console.log(error);
			}
			else
			{
				console.log("订阅了" + granted[0].topic);
			}
		});
	}

}

export function getMqtt(dotnetHelper, username, password, clientId)
{
	return new Mqtt(dotnetHelper, username, password, clientId);
}

File: Mqtt.razor.cs
using Microsoft.AspNetCore.Components;
using Microsoft.JSInterop;

namespace BlazorApp1.MqttComponent
{
	public partial class Mqtt
	{
		[Inject]
		IJSRuntime? JS { get; set; }

		//数据类型
		public class MqttOptions
		{
			public string Username { get; set; } = string.Empty;
			public string Password { get; set; } = string.Empty;
		}
		public class Msg
		{
			public string Topic { get; set; } = string.Empty;
			public byte[] Payload { get; set; } = Array.Empty<byte>();
		}

		//参数
		[Parameter]
		public MqttOptions? Options { get; set; }
		/// <summary>
		/// MQTT组件的默认配置
		/// </summary>
		public static MqttOptions DefaultOptions
		{
			get
			{
				return _defaultOptions;
			}
			set
			{
				_defaultOptions = value;
			}
		}
		static MqttOptions _defaultOptions = new();

		//事件
		[Parameter]
		public EventCallback OnConnectedCallback { get; set; }
		[Parameter]
		public EventCallback<Msg> OnReceivedCallback { get; set; }
		[Parameter]
		public EventCallback OnConnectFailCallback { get; set; }

		//给 JS 调用的函数
		/// <summary>
		/// 连接成功
		/// </summary>
		[JSInvokable]
		public async void OnConnected()
		{
			await OnConnectedCallback.InvokeAsync();
		}
		/// <summary>
		/// 接收到 mqtt 消息
		/// </summary>
		/// <param name="topic"></param>
		/// <param name="payload"></param>
		[JSInvokable]
		public async void OnReceived(string topic, byte[] payload)
		{
			await OnReceivedCallback.InvokeAsync(new Msg()
			{
				Topic = topic,
				Payload = payload,
			});
		}
		/// <summary>
		/// 连接失败
		/// </summary>
		[JSInvokable]
		public async void OnConnectFail()
		{
			await OnConnectFailCallback.InvokeAsync();
		}

		//提供给父组件的方法
		/// <summary>
		/// 发布消息
		/// </summary>
		/// <param name="topic"></param>
		/// <param name="payload"></param>
		public async void Publish(string topic, byte[] payload)
		{
			if (_mqtt is not null)
			{
				await _mqtt.InvokeVoidAsync("publish", topic, payload);
			}
		}
		/// <summary>
		/// 订阅主题
		/// </summary>
		/// <param name="topic"></param>
		public async void Subscribe(string topic)
		{
			if (_mqtt is not null)
			{
				await _mqtt.InvokeVoidAsync("subscribe", topic);
			}
		}
		/// <summary>
		/// 尝试连接
		/// </summary>
		public async void TryConnect()
		{
			string clientId = DefaultOptions.Username + "-" + Guid.NewGuid().ToString();
			//等待安装成功
			var tryConnect = async Task<bool> () =>
			  {
				  try
				  {
					  if (module != null)
					  {
						  Console.WriteLine("moudle不为空，尝试连接");
						  //如果用户没有设置 Options 则使用默认的设置
						  if (Options == null)
						  {
							  _mqtt = await module.InvokeAsync<IJSObjectReference>("getMqtt", _dotnetHelper, DefaultOptions.Username, DefaultOptions.Password, clientId);
						  }
						  else
						  {
							  _mqtt = await module.InvokeAsync<IJSObjectReference>("getMqtt", _dotnetHelper, Options.Username, Options.Password, clientId);
						  }
						  return true;
					  }
					  else
					  {
						  Console.WriteLine("moudle为空");
						  return false;
					  }
				  }
				  catch
				  {
					  Console.WriteLine("异常");
					  return false;
				  }
			  };
			while (!await tryConnect())
			{
				await Task.Delay(1000);
			}

		}

		//生命周期
		IJSObjectReference? module;
		IJSObjectReference? _mqtt;
		DotNetObjectReference<Mqtt>? _dotnetHelper;
		/// <summary>
		/// 组件初始化
		/// </summary>
		/// <param name="firstRender"></param>
		/// <returns></returns>
		protected override async Task OnAfterRenderAsync(bool firstRender)
		{
			if (firstRender)
			{
				while (JS == null)
				{
					await Task.Delay(1000);
				}
				Console.WriteLine("MQTT组件已加载");
				module = await JS.InvokeAsync<IJSObjectReference>("import", "./MqttComponent/Mqtt.razor.js");
				_dotnetHelper = DotNetObjectReference.Create(this);
				TryConnect();
			}
		}

	}
}
